# Exploratory Analysis - Stack Overflow Developer Survey (2023-2025)

## Objective
- Explore the cleaned dataset before answering the main question (check the README)
- Look for outliers, missing data and basic patterns by role and year

In [1]:
import pandas as pd

# Load cleaned data - Already unified across 3 years
df = pd.read_csv("../data/processed/survey_clean_2023_2025.csv")
df.head()

C:\Users\ArielOrtega\AppData\Local\Temp\ipykernel_15564\524532132.py:4: DtypeWarning: Columns (0: YearsCode, 1: AIComplex, 2: AIAgents) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("../data/processed/survey_clean_2023_2025.csv")


,MainBranch,Age,Employment,EdLevel,DevType,RoleClean,OrgSize,ICorPM,RemoteWork,Country,...,WorkExp,Currency,CompTotal,ConvertedCompYearly,AISelect,AISent,AIAcc,Year,AIComplex,AIAgents
0,None of these,18-24 years old,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2023,NaN,NaN
1,I am a developer by profession,25-34 years old,"Employed, full-time","Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Senior Executive (C-Suite, VP, etc.)",NaN,2 to 9 employees,People manager,Remote,United States of America,...,10.0,USD\tUnited States dollar,285000.0,285000.0,Yes,Indifferent,Other (please explain),2023,NaN,NaN
2,I am a developer by profession,45-54 years old,"Employed, full-time","Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Developer, back-end",NaN,"5,000 to 9,999 employees",Individual contributor,"Hybrid (some remote, some in-person)",United States of America,...,23.0,USD\tUnited States dollar,250000.0,250000.0,"No, and I don't plan to",NaN,NaN,2023,NaN,NaN
3,I am a developer by profession,25-34 years old,"Employed, full-time","Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Developer, front-end",NaN,100 to 499 employees,Individual contributor,"Hybrid (some remote, some in-person)",United States of America,...,7.0,USD\tUnited States dollar,156000.0,156000.0,"No, and I don't plan to",NaN,NaN,2023,NaN,NaN
4,I am a developer by profession,25-34 years old,"Employed, full-time;Independent contractor, fr...","Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Developer, full-stack",NaN,20 to 99 employees,Individual contributor,Remote,Philippines,...,6.0,PHP\tPhilippine peso,1320000.0,23456.0,Yes,Very favorable,Increase productivity;Greater efficiency;Speed...,2023,NaN,NaN


## 1.- Data Quality Check

In [2]:
# Check which columns have mixed types
for col in df.columns:
    types = df[col].dropna().apply(type).unique()
    if len(types) > 1:
        print(col, types)

YearsCode [<class 'str'> <class 'float'>]


### YearsCode - Mixed Formats

Found while exploring: 
- Numbers as text
- Numbers as decimals
- Two special values ("Less than 1 year" and "More than 50 years").

Needs cleaning before we can chart this column.

In [3]:
# Check unique YearsCode values with type
print(df["YearsCode"].unique())

[nan '18' '27' '12' '6' '21' '4' '5' '20' '14' '10' '15' '11' '3' '24' '8'
 '13' 'Less than 1 year' '16' '33' '22' '30' '32' '7' '35' '28' '40' '17'
 '29' '19' 'More than 50 years' '9' '38' '26' '34' '25' '2' '45' '23' '31'
 '43' '1' '48' '41' '50' '39' '42' '37' '36' '44' '46' '49' '47' '14.0'
 '10.0' '12.0' '5.0' '22.0' '20.0' '13.0' '30.0' '15.0' '9.0' '6.0' '3.0'
 '40.0' '8.0' '45.0' '35.0' '11.0' '7.0' '47.0' '41.0' '36.0' '23.0'
 '29.0' '26.0' '25.0' '38.0' '21.0' '19.0' '18.0' '16.0' '4.0' '17.0'
 '42.0' '27.0' '24.0' '2.0' '28.0' '37.0' '32.0' '1.0' '39.0' '53.0'
 '33.0' '50.0' '48.0' '44.0' '31.0' '46.0' '43.0' '49.0' '52.0' '34.0'
 '51.0' '58.0' '57.0' '54.0' '63.0' '60.0' '55.0' '65.0' '100.0' '99.0'
 '56.0' '69.0' 3.0 15.0 20.0 2.0 7.0 1.0 6.0 5.0 27.0 39.0 33.0 30.0 8.0
 22.0 35.0 18.0 10.0 21.0 25.0 12.0 32.0 40.0 28.0 9.0 11.0 4.0 36.0 29.0
 16.0 13.0 50.0 53.0 43.0 26.0 19.0 14.0 38.0 44.0 47.0 24.0 17.0 23.0
 41.0 34.0 42.0 45.0 37.0 46.0 31.0 48.0 55.0 60.0 49.0 100.0

In [4]:
# Clean YearsCode - Unify text and mixed number formats
df["YearsCode"] = df["YearsCode"].replace({
    "Less than 1 year": 0,
    "More than 50 years": 51
})
df["YearsCode"] = pd.to_numeric(df["YearsCode"], errors="coerce")

print(df["YearsCode"].describe())

count    190346.000000
mean         14.628230
std          10.926889
min           0.000000
25%           6.000000
50%          11.000000
75%          20.000000
max         100.000000
Name: YearsCode, dtype: float64


In [5]:
# Check how many responses claim 40+ years coding
print((df["YearsCode"] > 40).sum())

5557


In [6]:
# Check how many responses reach more than 60 years (probably bad data, not just long careers)
print((df["YearsCode"] > 60).sum())

71


#### Findings - YearsCode
- Found 71 responses reaching more than 60 years coding - Probably bad data (not real careers)

We should **remove** it before chart.

In [7]:
# YearsCode > 60 is not physically possible given the survey's max age bracket
# Set to NaN instead of dropping the row - Keeps the rest of that person's data
df.loc[df["YearsCode"] > 60, "YearsCode"] = pd.NA

print(df["YearsCode"].describe())

count    190275.000000
mean         14.600657
std          10.831002
min           0.000000
25%           6.000000
50%          11.000000
75%          20.000000
max          60.000000
Name: YearsCode, dtype: float64


### ConvertedCompYearly - Check for Outliers
Salary data in surveys is usually messy because people put the number in the wrong currency or write extra zeros by mistake.


In [8]:
print(df["ConvertedCompYearly"].describe())

count    9.540100e+04
mean     9.860668e+04
std      5.439221e+05
min      1.000000e+00
25%      4.006100e+04
50%      7.282100e+04
75%      1.196310e+05
max      7.435143e+07
Name: ConvertedCompYearly, dtype: float64


In [9]:
# Check extreme values on both ends
print((df["ConvertedCompYearly"] > 1_000_000).sum())
print((df["ConvertedCompYearly"] < 1_000).sum())

147
1953


In [10]:
# Check countries with high salary responses
print(df[df["ConvertedCompYearly"] > 1_000_000]["Country"].value_counts().head(10))

Country
United States of America                                45
Brazil                                                  10
United Kingdom of Great Britain and Northern Ireland     9
Germany                                                  7
Canada                                                   6
India                                                    6
Iran, Islamic Republic of...                             5
Australia                                                5
Ukraine                                                  5
Nigeria                                                  4
Name: count, dtype: int64


In [11]:
# Limit salary outliers - Values above one million could be currency or typo errors (not real salaries)
df.loc[df["ConvertedCompYearly"] > 1_000_000, "ConvertedCompYearly"] = pd.NA

print(df["ConvertedCompYearly"].describe())

count      95254.000000
mean       89853.425735
std        80449.315603
min            1.000000
25%        40000.000000
50%        72818.000000
75%       118987.000000
max      1000000.000000
Name: ConvertedCompYearly, dtype: float64


#### Findings - High salary check

**Found:** 147 responses over one million dollars per year

**Checked:** by country

**Pattern:** No economic reason for these amounts - They are spread across countries where one million dollars salaries are not realistic.

**Conclusion:** Probably currency or typo errors - Set NaN

In [12]:
# Check countries with low salary responses
print(df[df["ConvertedCompYearly"] < 1000]["Country"].value_counts().head(10))

Country
Ukraine                     284
United States of America    136
India                        97
Nigeria                      88
Pakistan                     77
Colombia                     67
Indonesia                    60
Brazil                       56
Turkey                       50
Argentina                    47
Name: count, dtype: int64


In [13]:
# Check how low the low salaries are
print(df[df["ConvertedCompYearly"] < 1000]["ConvertedCompYearly"].describe())

count    1953.000000
mean      306.359959
std       300.103915
min         1.000000
25%        48.000000
50%       182.000000
75%       537.000000
max       999.000000
Name: ConvertedCompYearly, dtype: float64


In [14]:
# Check how many fall below a realistic minimum (100 dollars per year)
print((df["ConvertedCompYearly"] < 100).sum())

720


In [15]:
# Below 100 dollars per year is not a realistic salary anywhere
df.loc[df["ConvertedCompYearly"] < 100, "ConvertedCompYearly"] = pd.NA

print(df["ConvertedCompYearly"].describe())

count      94534.000000
mean       90537.517782
std        80370.846559
min          100.000000
25%        40814.000000
50%        73375.000000
75%       120000.000000
max      1000000.000000
Name: ConvertedCompYearly, dtype: float64


#### Findings - Low salary check:

**Found:** 1953 responses under $1000 per year

**Checked:** by country - Low cost of living countries
(Ukraine, India, Nigeria, Pakistan)

**Looked closer:** 720 of those are under $100 per year - There is not any country with real salary as low as that

**Conclusion:** 
- Below 100 dollars set NaN (bad data)
- Between 100 - 1000 dollars kept as real salaries.

### WorkExp - Check for Outliers

Same risk as YearsCode - Open number field (Check just in case)

In [16]:
print(df["WorkExp"].describe())

count    116130.000000
mean         12.145690
std           9.806774
min           0.000000
25%           5.000000
50%          10.000000
75%          17.000000
max         100.000000
Name: WorkExp, dtype: float64


In [17]:
# Check how many reach more than 60 years of work experience
print((df["WorkExp"] > 60).sum())

54


In [18]:
# WorkExp > 60 is not physically possible - Set NaN
df.loc[df["WorkExp"] > 60, "WorkExp"] = pd.NA

print(df["WorkExp"].describe())

count    116076.000000
mean         12.108868
std           9.654321
min           0.000000
25%           5.000000
50%          10.000000
75%          17.000000
max          60.000000
Name: WorkExp, dtype: float64


#### Findings - WorkExp:
54 responses over 60 years work experience - Not physically possible (set NaN)

## 2.- Missing Data Check

**How complete is each column?** - It helps to decide what is usable for charts later.

In [19]:
print(df.isnull().sum())

MainBranch                  0
Age                         0
Employment               2138
EdLevel                  6906
DevType                 23815
RoleClean              195176
OrgSize                 57111
ICorPM                  97100
RemoteWork              41416
Country                 21472
Industry               104538
YearsCode               13537
WorkExp                 87736
Currency                56357
CompTotal               96981
ConvertedCompYearly    109278
AISelect                21212
AISent                  62971
AIAcc                   94619
Year                        0
AIComplex              133508
AIAgents               171893
dtype: int64


In [20]:
# Check missing AI data - Only in the 4 roles that we are interested for the main question
role_subset = df[df["RoleClean"].notna()]
print(role_subset.shape)
print(role_subset[["AISelect", "AISent", "AIAcc"]].isnull().sum())

(8636, 22)
AISelect     505
AISent      2093
AIAcc       3583
dtype: int64


#### Findings - Missing Data
In the 4 roles: 
- `AISelect` is well answered (505 missing).
- `AISent` and `AIAcc` drop more (2093 and 3583 missing) - Expect small samples ( Splitting by role and year)

## 3.- Main Question - AI Adoption by Role

**First look:** AISelect, AISent and AIAcc by role across the 3 years.

### AISelect - Uses AI?

In [21]:
# AISelect by role and year - Uses AI?
print(role_subset.groupby(["RoleClean", "Year"])["AISelect"].value_counts(normalize=True)) # Percentages

RoleClean               Year  AISelect                                   
Data Analyst            2023  Yes                                            0.416965
                              No, but I plan to soon                         0.295102
                              No, and I don't plan to                        0.287933
                        2024  Yes                                            0.571705
                              No, and I don't plan to                        0.242248
                              No, but I plan to soon                         0.186047
                        2025  Yes, I use AI tools daily                      0.362205
                              Yes, I use AI tools weekly                     0.216535
                              No, and I don't plan to                        0.165354
                              No, but I plan to soon                         0.137795
                              Yes, I use AI tools monthly or infre

#### Findings - AISelect

**Found:** Data Analyst behind Data Engineer and Data Scientist - daily use

**2025 daily use:** Data Analyst 36% - Data Engineer 52% - Data Scientist 52%

**Database Administrator:** last place (24% daily)

### AISent - How favorable toward AI?

In [22]:
# AISent by role and year - How favorable toward AI?
print(role_subset.groupby(["RoleClean", "Year"])["AISent"].value_counts(normalize=True))

RoleClean               Year  AISent          
Data Analyst            2023  Favorable           0.447811
                              Very favorable      0.355219
                              Indifferent         0.127946
                              Unfavorable         0.035354
                              Unsure              0.026936
                                                    ...   
Database Administrator  2025  Indifferent         0.235294
                              Very favorable      0.213235
                              Unfavorable         0.110294
                              Very unfavorable    0.088235
                              Unsure              0.073529
Name: proportion, Length: 71, dtype: float64


In [23]:
# % favorable or very favorable toward AI by role and year
favorable = role_subset[role_subset["AISent"].isin(["Favorable", "Very favorable"])]
print(favorable.groupby(["RoleClean", "Year"]).size() / role_subset.groupby(["RoleClean", "Year"])["AISent"].count())

RoleClean               Year
Data Analyst            2023    0.803030
                        2024    0.731458
                        2025    0.632411
Data Engineer           2023    0.793333
                        2024    0.752096
                        2025    0.676271
Data Scientist          2023    0.824942
                        2024    0.813478
                        2025    0.661470
Database Administrator  2023    0.625767
                        2024    0.672727
                        2025    0.492647
dtype: float64


#### Findings - AISent
**Found:** Favorable/very favorable drops from 2023  to 2025 - All 4 roles

**Data Analyst:** From 80% to 63%

**Pattern:** Same drop across all roles - Not just one

### AIAcc - Trust in AI Accuracy?

In [24]:
# % Who trust AI accuracy by role and year - Highly trust or Somewhat trust
trust = role_subset[role_subset["AIAcc"].isin(["Highly trust", "Somewhat trust"])]
print(trust.groupby(["RoleClean", "Year"]).size() / role_subset.groupby(["RoleClean", "Year"])["AIAcc"].count())

RoleClean               Year
Data Analyst            2023         NaN
                        2024    0.552901
                        2025    0.440476
Data Engineer           2023         NaN
                        2024    0.453577
                        2025    0.393888
Data Scientist          2023         NaN
                        2024    0.506241
                        2025    0.388393
Database Administrator  2023         NaN
                        2024    0.614286
                        2025    0.313433
dtype: float64


#### Findings - AIAcc

**Found:** Trust in AI accuracy - Drop pattern as AISent

**Data Analyst:** 2024 (55%) - 2025 (44%)

**Database Administrator:** Biggest drop - From 61% to 31%

### AIComplex - Trust for Complex Tasks?

In [32]:
# Check AIComplex categories first
print(role_subset["AIComplex"].unique())

<StringArray>
[                                                    nan,
                         'Bad at handling complex tasks',
         'Good, but not great at handling complex tasks',
                   'Very well at handling complex tasks',
         'Neither good or bad at handling complex tasks',
                   'Very poor at handling complex tasks',
 'I don't use AI tools for complex tasks / I don't know']
Length: 7, dtype: str


In [27]:
# % Who rate AI as good or very well for complex tasks by role and year
complex_trust = role_subset[role_subset["AIComplex"].isin([
    "Very well at handling complex tasks",
    "Good, but not great at handling complex tasks"
])]
print(complex_trust.groupby(["RoleClean", "Year"]).size() / role_subset.groupby(["RoleClean", "Year"])["AIComplex"].count())

RoleClean               Year
Data Analyst            2023         NaN
                        2024    0.468966
                        2025    0.351779
Data Engineer           2023         NaN
                        2024    0.383792
                        2025    0.321976
Data Scientist          2023         NaN
                        2024    0.426982
                        2025    0.395973
Database Administrator  2023         NaN
                        2024    0.571429
                        2025    0.238806
dtype: float64


#### Findings - AIComplex

**Found:** Trust drops from 2024  to 2025 - All 4 roles

**Data Analyst:** From 47% to 35%

**Risk:** DBA drop from 57% to 24% - Small sample (n=70 in 2024)

### AIAgents - Using AI Agents?

In [33]:
# Check AIAgents categories first
print(role_subset["AIAgents"].unique())

<StringArray>
[                                                    nan,
                               'No, and I don't plan to',
                                     'No, but I plan to',
                   'Yes, I use AI agents at work weekly',
                    'Yes, I use AI agents at work daily',
  'Yes, I use AI agents at work monthly or infrequently',
 'No, I use AI exclusively in copilot/autocomplete mode']
Length: 7, dtype: str


In [35]:
# % Using AI agents (daily, weekly or monthly) by role - 2025 only
using_agents = role_subset[role_subset["AIAgents"].isin([
    "Yes, I use AI agents at work daily",
    "Yes, I use AI agents at work weekly",
    "Yes, I use AI agents at work monthly or infrequently"
])]
print(using_agents.groupby("RoleClean").size() / role_subset.groupby("RoleClean")["AIAgents"].count())

RoleClean
Data Analyst              0.312500
Data Engineer             0.298214
Data Scientist            0.312354
Database Administrator    0.222222
dtype: float64


#### Findings - AIAgents

**Found:** Data Analyst 31%  -  Data Scientist 31%

**Pattern:** Different from AIComplex and AISent - Data analyst followed both of them.

**Note:** Data Engineer close (30%) - Database Administrator last (22%)